vpr/src/place/annealer.cpp

```diff
@@ -0,0 +1,898 @@
+
+#include "annealer.h"
+
+#include <algorithm>
+#include <cmath>
+
+#include "globals.h"
+#include "draw_global.h"
+#include "vpr_types.h"
+#include "place_util.h"
+#include "placer_state.h"
+#include "move_utils.h"
+#include "manual_move_generator.h"
+#include "noc_place_utils.h"
+#include "NetPinTimingInvalidator.h"
+#include "place_timing_update.h"
+#include "read_place.h"
+#include "placer_breakpoint.h"
+
+/**
+ * @brief Check if the setup slack has gotten better or worse due to block swap.
+ *
+ * Get all the modified slack values via the PlacerSetupSlacks class, and compare
+ * then with the original values at these connections. Sort them and compare them
+ * one by one, and return the difference of the first different pair.
+ *
+ * If the new slack value is larger(better), than return a negative value so that
+ * the move will be accepted. If the new slack value is smaller(worse), return a
+ * positive value so that the move will be rejected.
+ *
+ * If no slack values have changed, then return an arbitrary positive number. A
+ * move resulting in no change in the slack values should probably be unnecessary.
+ *
+ * The sorting is need to prevent in the unlikely circumstances that a bad slack
+ * value suddenly got very good due to the block move, while a good slack value
+ * got very bad, perhaps even worse than the original worse slack value.
+ */
+static float analyze_setup_slack_cost(const PlacerSetupSlacks* setup_slacks,
+                                      const PlacerState& placer_state);
+
+
```

Comment:
Maybe put a separator betweeen local function declarations and function definitions, e.g.

/************************** Function definitions **********************/

(It looks funny to have a declaration followed immediately by the definition otherwise).